### PR TITLE
host doc on s3

### PR DESCRIPTION
**Pull request by szha**
## Description ##
host the doc built by CI for each PR on s3, as part of deploy step

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] ~All changes have test coverage:~ testing here in PR, as the change is in Jenkins
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] host doc for PR in s3

### CODE REVIEW

File: tests/ci_build/deploy/ci_deploy_doc.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#
+# Execute command within a docker container
+#
+# Usage: ci_deploy_doc.sh <PR_ID> <BUILD_ID>
+#
+# PR_ID: the PR number
+#
+# BUILD_ID: the current build ID for the specified PR
+#
+aws s3 sync --delete docs/_build/html/ s3://mxnet-ci-doc/$1/$2 --acl public-read \
```
Comment (marcoabreu) [inline tests/ci_build/deploy/ci_deploy_doc.sh]: "acl public-read". Could a PR modify this flag in order to enable public-write? Could be problematic.

Would it rather be possible to set permissions on a bucket-level that bubble down to every object?

Comment (szha) [reply tests/ci_build/deploy/ci_deploy_doc.sh]: Yes, I will need to open up public-read for the bucket. The problem remains that changes to this script takes effect for anyone. The alternative is to put this command in Jenkinsfile, in which only committers' changes can take effect immediately.

File: Jenkinsfile [hunk 2]
```diff
@@ -692,6 +692,7 @@ try {
         init_git()
         sh "make clean"
         sh "make docs"
+        sh tests/ci_build/deploy/ci_deploy_doc.sh ${env.BRANCH_NAME} ${env.CHANGE_ID}
```
Comment (marcoabreu) [inline Jenkinsfile]: Add quotation marks

File: tests/ci_build/deploy/ci_deploy_doc.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+#
+# Execute command within a docker container
```
Comment (marcoabreu) [inline tests/ci_build/deploy/ci_deploy_doc.sh]: "outside a docker container". AWS CLI is not available in docker.

**Comment by szha** [comment]
Now the doc is hosted at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-9928/25/index.html

File: tests/ci_build/deploy/aws [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env python
```
Comment (marcoabreu) [inline tests/ci_build/deploy/aws]: This file has no type (.py)

**Comment by marcoabreu** [comment]
Awesome, great job sheng!

File: tests/ci_build/deploy/aws (hunk 4 above)
Comment (szha) [reply tests/ci_build/deploy/aws]: This is a copy of the driver script for awscli, as the installation in user space doesn't include it. It's usually at `/usr/local/bin/aws` but slaves don't have write access to this path.

File: tests/ci_build/deploy/ci_deploy_doc.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+# Execute command outside a docker container
+#
+# Usage: ci_deploy_doc.sh <PR_ID> <BUILD_ID>
+#
+# PR_ID: the PR number
+#
+# BUILD_ID: the current build ID for the specified PR
+#
+
+# TODO szha@: installation of awscli here should be removed once slave hosts have them during
```
Comment (eric-haibin-lin) [inline tests/ci_build/deploy/ci_deploy_doc.sh]: Ideally we should leave no TODOs in the code. Maybe a github issue to track?

Comment (marcoabreu) [reply tests/ci_build/deploy/ci_deploy_doc.sh]: This is already implemented in the test environment and just waits for redeployment

**Pull request merged**